mono/mini/aot-mangler.c

```diff
@@ -0,0 +1,1244 @@
+	}
+	if (!strcmp (in [i], "method_inst")) {
+		mono_json_writer_indent (out);
+		mono_json_writer_object_key(out, "method_inst");
+		mono_json_writer_array_begin (out);
+		i++;
+		while (strcmp (in [i], "]")) {
+			mono_json_writer_indent (out);
+			mono_json_writer_printf (out, "\"%s\",\n", in [i++]);
+		}
+		mono_json_writer_indent (out);
+		mono_json_writer_array_end (out);
+		mono_json_writer_printf (out, "\n");
+	}
+
+	demangle_assert(error, !strcmp (in [i], "]"));
+	if (!mono_error_ok (error))
+		goto fail;
+
+	mono_json_writer_indent (out);
+	mono_json_writer_object_end (out);
+	return i + 1;
+
+fail:
+	return -1;
+}
+
+/*
+ * decribe_mangled_klass: 
+ *   Turn the tokens of a mangled class into a JSON object
+ *
+ * \param out JsonWriter that has been initialized. This emits a json Object, complete with enclosing braces
+ * \param in Array of tokens of mangled string
+ * \param len Length of array of tokens
+ * \param error The error state
+ *
+ * \returns number of tokens consumed, -1 on failure.
+ */
+static int
+describe_mangled_klass (JsonWriter *out, gchar **in, size_t len, MonoError *error)
+{
+	int i = 0;
+	error_init (error);
+	demangle_assert(error, !strcmp (in [i++], "["));
+	demangle_assert(error, !strcmp (in [i++], "klass"));
+	demangle_assert(error, !strcmp (in [i++], "namespace:"));
+	if (!mono_error_ok (error))
+		goto fail;
+
+	mono_json_writer_object_begin (out);
+
+	mono_json_writer_indent (out);
+	mono_json_writer_object_key (out, "category:");
+	mono_json_writer_printf (out, "\"class\",\n");
+
+	mono_json_writer_indent (out);
+	mono_json_writer_object_key(out, "namespace");
+	mono_json_writer_printf (out, "\"");
+
+	while (strcmp (in [i], "name:"))
+		mono_json_writer_printf (out, "%s", in [i++]);
+	mono_json_writer_printf (out, "\",\n");
+
+	demangle_assert(error, !strcmp (in [i], "name:"));
+	if (!mono_error_ok (error))
+		goto fail;
+
+	mono_json_writer_indent (out);
+	mono_json_writer_object_key(out, "name");
+	mono_json_writer_printf (out, "\"");
+	i++;
+
+	while (strcmp (in [i], "]"))
+		mono_json_writer_printf (out, "%s", in [i++]);
+	mono_json_writer_printf (out, "\",\n");
+
+	demangle_assert(error, !strcmp (in [i], "]"));
+	if (!mono_error_ok (error))
+		goto fail;
+
+	mono_json_writer_indent_pop (out);
+	mono_json_writer_indent (out);
+	mono_json_writer_object_end (out);
+
+	return i + 1;
+
+fail:
+	return -1;
+}
+
+/*
+ * decribe_mangled_method: 
+ *   Turn the tokens of a mangled method into a JSON object
+ *
+ * \param out JsonWriter that has been initialized. This emits a json Object, complete with enclosing braces
+ * \param in Array of tokens of mangled string
+ * \param len Length of array of tokens
+ * \param error The error state
+ *
+ * \returns number of tokens consumed, -1 on failure.
+ */
+static int
+describe_mangled_method (JsonWriter *out, gchar **in, size_t len, MonoError *error)
+{
+	error_init (error);
+	if (len > 2 && !strcmp (in [0], "[") && !strcmp (in [1], "wrapper"))
```

Comment:
What having '[' as the first character means?
Mangled names can't have '[' in then.